### PR TITLE
feat(grpc status codes): report grpc status codes via mixer

### PR DESCRIPTION
**Pull request by douglas-reid**
This PR implements the changes approved in https://tinyurl.com/yy2cwmv5 to report gRPC status codes in metrics (and provide better alignment with logs and metrics when reporting protocol). It also cleans up unused metrics dimensions.

Other PRs will be required to support the same feature set in v2 telemetry and to update the operator/installer config templates.

Partial fix for https://github.com/istio/istio/issues/12846

[ ] Configuration Infrastructure
[ ] Docs
[ ] Installation
[ ] Networking
[ ] Performance and Scalability
[ X ] Policies and Telemetry
[ ] Security
[ ] Test and Release
[ ] User Experience
[ ] Developer Infrastructure

### CODE REVIEW

File: install/kubernetes/helm/istio/charts/mixer/templates/config.yaml [hunk 1]
```diff
@@ -430,9 +429,8 @@ spec:
       destination_service_namespace: destination.service.namespace | "unknown"
       request_protocol: api.protocol | context.protocol | "unknown"
       response_code: response.code | 200
+      grpc_response_status: response.grpc_status | 0
       response_flags: context.proxy_error_code | "-"
-      permissive_response_code: rbac.permissive.response_code | "none"
-      permissive_response_policyid: rbac.permissive.effective_policy_id | "none"
```
Comment (kyessenov) [inline install/kubernetes/helm/istio/charts/mixer/templates/config.yaml]: why are these removed? cardinality issue?

File: install/kubernetes/helm/istio/charts/mixer/templates/config.yaml [hunk 2]
```diff
@@ -467,9 +465,8 @@ spec:
       destination_service_namespace: destination.service.namespace | "unknown"
       request_protocol: api.protocol | context.protocol | "unknown"
       response_code: response.code | 200
+      grpc_response_status: response.grpc_status | 0
```
Comment (kyessenov) [inline install/kubernetes/helm/istio/charts/mixer/templates/config.yaml]: is the status an int or a string? I think it's a string (by mistake)

Comment (douglas-reid) [reply install/kubernetes/helm/istio/charts/mixer/templates/config.yaml]: apparently it is a string (based on the test failures).  Is "0" the right default, or "unknown" ?

File: install/kubernetes/helm/istio/charts/mixer/templates/config.yaml (hunk 1 above)
Comment (douglas-reid) [reply install/kubernetes/helm/istio/charts/mixer/templates/config.yaml]: For months I've been asking if these are used, and have yet to find anyone that is actually using these dimensions. I *thought* we were not carrying them forward into v2 (though the code now says otherwise).

@mandarjog objections to removing?

Comment (kyessenov) [reply install/kubernetes/helm/istio/charts/mixer/templates/config.yaml]: @liminw @yangminzhu can we remove these?

File: install/kubernetes/helm/istio/charts/mixer/templates/config.yaml (hunk 2 above)
Comment (kyessenov) [reply install/kubernetes/helm/istio/charts/mixer/templates/config.yaml]: it's set if grpc status header is present, and not set at all otherwise (see `
`src/envoy/http/mixer/report_data.h`)

Comment (douglas-reid) [reply install/kubernetes/helm/istio/charts/mixer/templates/config.yaml]: right, but do we want the unset to be represented in metrics as "0" (like we default to "200" for http) or "unknown" ?

Comment (kyessenov) [reply install/kubernetes/helm/istio/charts/mixer/templates/config.yaml]: I think empty is better. It shouldn't really be unknown for grpc ever.

**Comment by douglas-reid** [comment]
Ping. This now includes updates for the merged `manifests/`.

**Comment by howardjohn** [comment]
i am not qualified to review, lmk once a mixer owner approves and I can give it the approval

File: install/kubernetes/helm/istio/charts/mixer/templates/config.yaml (hunk 1 above)
Comment (jshaughn) [reply install/kubernetes/helm/istio/charts/mixer/templates/config.yaml]: fwiw, kialki does not use these.

File: install/kubernetes/helm/istio/charts/mixer/templates/config.yaml (hunk 2 above)
Comment (jshaughn) [reply install/kubernetes/helm/istio/charts/mixer/templates/config.yaml]: So just to confirm, the decision to have separate response_code and grpc_response_status fields is because even if request_protocol=grpc, response.code is still set?

Also, I think 0 is good for unset, just like the 200.  I like a default being the same type as an actual value.

File: install/kubernetes/helm/istio/charts/mixer/templates/config.yaml (hunk 1 above)
Comment (mandarjog) [reply install/kubernetes/helm/istio/charts/mixer/templates/config.yaml]: Yes please remove, @yangminzhu last chance to object.

Comment (yangminzhu) [reply install/kubernetes/helm/istio/charts/mixer/templates/config.yaml]: Sorry I missed this PR. It should be fine to remove because the new authorization policy doesn't support the permissive mode but I'm not sure if it's used by anyone in the old RBAC policy for now. cc @liminw

**Comment by jshaughn** [comment]
@douglas-reid Will this make it into 1.5?

**Comment by douglas-reid** [comment]
@jshaughn i hope so. i'm certainly trying to make it happen.

**Comment by douglas-reid** [comment]
@jshaughn the biggest issue is making sure we have equivalent functionality with v2.

File: install/kubernetes/helm/istio/charts/mixer/templates/config.yaml (hunk 2 above)
Comment (jshaughn) [reply install/kubernetes/helm/istio/charts/mixer/templates/config.yaml]: @douglas-reid I know the ship has sailed but perhaps the new field should have been "request_protocol_status" or something like that.  And contained the response status for the specified request_protocol.  Basically replacing request_code, which has now been relegated to effectively being "http_code".  So existing consumer code could still assume http codes in request_code and write new code using the new field, which could support the two existing protocols and if necessary, additional protocols somewhere down the line.  And "response_code" could have been deprecated.

Comment (douglas-reid) [reply install/kubernetes/helm/istio/charts/mixer/templates/config.yaml]: @jshaughn perhaps -- that's closer to my original desire. I think we can rethink that for 1.6/1.7 when maybe we do an overhaul of the metrics (and they are controlled all via v2). however, at this point, i'd like to get the approved design implemented and into istio/istio first.

Comment (jshaughn) [reply install/kubernetes/helm/istio/charts/mixer/templates/config.yaml]: @douglas-reid yup, totally agree.  And I've been coding kiali the past days in anticipation of the new grpc_request_status field, and to be back compatible.

**Comment by douglas-reid** [comment]
Gentle ping on a review here. This PR brings the v1 telemetry pipeline in line with changes already made to v2.

**Review by mandarjog — APPROVED** [review]
This is one of the last features added Mixer.

**Review by howardjohn — APPROVED** [review]

**Review by kyessenov — APPROVED** [review]

**Comment by israel-hdez** [comment]
Perhaps, could you update the Default Metrics page? https://istio.io/docs/reference/config/policy-and-telemetry/metrics/

I also see on that page the sentence `They can be found here under the section with “kind: metric”` while, that `kind: metric` looks non existent in the current config.yaml.
